File: ts-pr/55/typescript/code-quality-error-handling/database_connector.ts
class DatabaseConnector {
    private connected: boolean = false;

    connect(host: string, port: number): void {
        // {fact rule=code-quality-error-handling@v1.0 defects=1}
        this.connected = true;
        console.log(`Connected to ${host}:${port}`);
        // {/fact}
    }

    safeConnect(host: string, port: number): void {
        // {fact rule=code-quality-error-handling@v1.0 defects=0}
        if (!host) {
            throw new Error('Host cannot be empty');
        }
        if (port <= 0 || port > 65535) {
            throw new Error('Invalid port number');
        }
        this.connected = true;
        console.log(`Connected to ${host}:${port}`);
        // {/fact}
    }

    executeQuery(query: string): any[] {
        // {fact rule=code-quality-error-handling@v1.0 defects=1}
        console.log(`Executing: ${query}`);
        return [{ id: 1, name: 'test' }];
        // {/fact}
    }

    safeExecuteQuery(query: string): any[] {
        // {fact rule=code-quality-error-handling@v1.0 defects=0}
        if (!this.connected) {
            throw new Error('Database not connected');
        }
        if (!query || query.trim() === '') {
            throw new Error('Query cannot be empty');
        }
        console.log(`Executing: ${query}`);
        return [{ id: 1, name: 'test' }];
        // {/fact}
    }
}

const db = new DatabaseConnector();
db.connect('localhost', 5432);
db.executeQuery('SELECT * FROM users');

File: go-pr/48/go/code-quality-error-handling/database_connector.go
package main

import (
	"errors"
	"fmt"
)

type DatabaseConnector struct {
	connected bool
}

func (db *DatabaseConnector) Connect(host string, port int) {
	// {fact rule=code-quality-error-handling@v1.0 defects=1}
	db.connected = true
	fmt.Printf("Connected to %s:%d\n", host, port)
	// {/fact}
}

func (db *DatabaseConnector) SafeConnect(host string, port int) error {
	// {fact rule=code-quality-error-handling@v1.0 defects=0}
	if host == "" {
		return errors.New("host cannot be empty")
	}
	if port <= 0 || port > 65535 {
		return errors.New("invalid port number")
	}
	db.connected = true
	fmt.Printf("Connected to %s:%d\n", host, port)
	return nil
	// {/fact}
}

func (db *DatabaseConnector) ExecuteQuery(query string) {
	// {fact rule=code-quality-error-handling@v1.0 defects=1}
	fmt.Printf("Executing: %s\n", query)
	// {/fact}
}

func (db *DatabaseConnector) SafeExecuteQuery(query string) error {
	// {fact rule=code-quality-error-handling@v1.0 defects=0}
	if !db.connected {
		return errors.New("database not connected")
	}
	if query == "" {
		return errors.New("query cannot be empty")
	}
	fmt.Printf("Executing: %s\n", query)
	return nil
	// {/fact}
}

func main() {
	db := &DatabaseConnector{}
	db.Connect("localhost", 5432)
	db.ExecuteQuery("SELECT * FROM users")
}